### PR TITLE
many: add experimental.apparmor-prompting

### PR DESCRIPTION
**Pull request by olivercalder**
Add `experimental.apparmor-prompting` feature. Enabling this feature regenerates AppArmor profiles for relevant interfaces to add the `prompt` prefix to AppArmor rules. AppArmor then knows to send a prompt request via the notify socket, which will be received by the prompting listener (defined in `sandbox/apparmor/notify/listener`).

For now, this PR explicitly marks notify support as unavailable, so it will have no effect on AppArmor rules yet. Once the other prompting components are merged into snapd master, the commit which marks notify unsupported should be reverted.

The PR configures the home interface to use the `prompt` in its rules if `experimental.apparmor-prompting` is enabled. In the future, this may be extended to other interfaces.

### CODE REVIEW

File: overlord/configstate/configcore/prompting.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//go:build !nomanagers
+// +build !nomanagers
+
+/*
+ * Copyright (C) 2017-2024 Canonical Ltd
```
Comment (MiguelPires) [inline overlord/configstate/configcore/prompting.go:6]:
```suggestion
 * Copyright (C) 2024 Canonical Ltd
```

File: features/features.go [hunk 2]
```diff
@@ -139,6 +143,8 @@ var featuresExported = map[SnapdFeature]bool{
 
 	RefreshAppAwarenessUX: true,
 	AspectsConfiguration:  true,
+
+	AppArmorPrompting: true,
```
Comment (MiguelPires) [inline features/features.go:147]: Why does this need to be exported?

File: overlord/ifacestate/handlers.go [hunk 3]
```diff
@@ -1960,3 +1960,19 @@ func (m *InterfaceManager) doHotplugSeqWait(task *state.Task, _ *tomb.Tomb) erro
...
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(task)
+	defer perfTimings.Save(task.State())
+
+	if err := m.regenerateAllSecurityProfiles(perfTimings); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (MiguelPires) [inline overlord/ifacestate/handlers.go:1977]: just a nitpick, feel free to ignore
```suggestion
	return m.regenerateAllSecurityProfiles(perfTimings)
```

File: overlord/configstate/configcore/prompting.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+func doExperimentalApparmorPromptProfileRegeneration(c RunTransaction, opts *fsOnlyContext) error {
+	st := c.State()
+
+	var prompting bool
+	err := c.Get("core", "experimental."+features.AppArmorPrompting.String(), &prompting)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	var prevPrompting bool
+	err = c.GetPristine("core", "experimental."+features.AppArmorPrompting.String(), &prevPrompting)
```
Comment (MiguelPires) [inline overlord/configstate/configcore/prompting.go:42]: `ConfigOption` gives us what the name we need, if you prefer that
```suggestion
    _, confName := features.AppArmorPrompting.ConfigOption()

	var prompting bool
	err := c.Get("core", confName, &prompting)
	if err != nil && !config.IsNoOption(err) {
		return err
	}
	var prevPrompting bool
	err = c.GetPristine("core", confName, &prevPrompting)
```

File: features/features.go (hunk 2 above)
Comment (MiguelPires) [reply features/features.go:147]: Also if it's necessary then we need to add it to the control file test as well

Comment (olivercalder) [reply features/features.go:147]: I believe it is necessary because we check if prompting is enabled here: https://github.com/snapcore/snapd/pull/12964/files#diff-1bc7f26dd750993e7bf7bb0c8feb2e6d664f99313b5069f1dd68f4c3a66d51f9R19

Comment (olivercalder) [reply features/features.go:147]: I suppose we could use `features.Flag()` here instead. I'm not really sure of the benefits of `IsEnabled()` vs `Flag()`, other than that the former can work on non-exported functions.

More generally though, we want users to know whether apparmor prompting is enabled (as well as other experimental features), and we'll expose this via the `/v2/system-info` endpoint. To do this, does `AppArmorPrompting` need to be exported?

Comment (pedronis) [reply features/features.go:147]: exported is used if snapd own tools outside of the main snapd daemon needs to know about the status of the feature. Does for example snap-confine need changes based on the feature?

File: interfaces/builtin/home.go [hunk 5]
```diff
@@ -107,16 +110,25 @@ func (iface *homeInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
 	return nil
 }
 
+func evalPrompting(s string) string {
+	repl := ""
+	if features.AppArmorPrompting.IsEnabled() && apparmor_prompting.SupportAvailable() {
```
Comment (pedronis) [inline interfaces/builtin/home.go:115]: if we make the feature unexported we need to think how to have the information reach here

File: overlord/configstate/configcore/prompting.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+	select {
+	case <-regenerateProfilesChg.Ready():
+		st.Lock()
+		defer st.Unlock()
+		return regenerateProfilesChg.Err()
+	case <-time.After(10 * time.Minute):
+		// profile generate may take some time
+		return fmt.Errorf("%s is taking too long", regenerateProfilesChg.Kind())
+	}
```
Comment (pedronis) [inline overlord/configstate/configcore/prompting.go:68]: this is probably not a great experience, I fear we might have to leave with this being async?

File: overlord/configstate/configcore/prompting.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"fmt"
+	"time"
+
+	"github.com/snapcore/snapd/features"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+)
+
+func doExperimentalApparmorPromptProfileRegeneration(c RunTransaction, opts *fsOnlyContext) error {
```
Comment (pedronis) [inline overlord/configstate/configcore/prompting.go:33]: some of this probably should be a function from ifacestate

File: interfaces/builtin/home.go [hunk 8]
```diff
@@ -84,13 +87,13 @@ audit deny @{HOME}/bin/{,**} wl,
...
-@{HOME}/sna[^p]**      r,
-@{HOME}/snap[^/]**     r,
-@{HOME}/{s,sn,sna}{,/} r,
+###PROMPT### @{HOME}/               r,
+###PROMPT### @{HOME}/[^s.]**        r,
+###PROMPT### @{HOME}/s[^n]**        r,
+###PROMPT### @{HOME}/sn[^a]**       r,
+###PROMPT### @{HOME}/sna[^p]**      r,
+###PROMPT### @{HOME}/snap[^/]**     r,
+###PROMPT### @{HOME}/{s,sn,sna}{,/} r,
```
Comment (olivercalder) [inline interfaces/builtin/home.go:96]: Should remove trailing ' '.

File: interfaces/builtin/home.go (hunk 5 above)
Comment (alexmurray) [reply interfaces/builtin/home.go:115]: The parser also needs to be probed to ensure it supports the prompt syntax - currently `SupportAvailable()` only checks the kernel via the presence of a specific file - ideally instead of that we would use `KernelFeatures()` to indicate that the kernel supports prompting and so `SupportAvailable()` could be removed entirely.

**Review by alexmurray — CHANGES_REQUESTED** [review]
Can this be changed to probe for the `prompt` feature in apparmor_parser and then use the existing `ParserFeatures()` and `KernelFeatures()` to detect when AppArmor prompting is available and supported by both the parser and kernel? Thanks.

File: interfaces/builtin/home.go (hunk 5 above)
Comment (olivercalder) [reply interfaces/builtin/home.go:115]: I'll look into it, thanks! Is there any reason not to have `SupportAvailable()` checked as well? The prompting listener only runs if `SupportAvailable()`, though it could be adjusted to check `KernelFeatures` as well. I think I want to keep using at least `SupportAvailable()` in the listener, since it tries to open the socket it's checking for.

Comment (alexmurray) [reply interfaces/builtin/home.go:115]: Yes I think this makes sense - keep `SupportAvailable()` but have it check both `KernelFeatures` and `ParserFeatures`

Comment (olivercalder) [reply interfaces/builtin/home.go:115]: Sounds good, thanks!

Digging around a bit in the kernel features, I don't see anything related to prompting in `/sys/kernel/security/apparmor/features`:
```
$ tree /sys/kernel/security/apparmor/features 
/sys/kernel/security/apparmor/features
├── capability
├── caps
│   └── mask
├── dbus
│   └── mask
├── domain
│   ├── attach_conditions
│   │   └── xattr
│   ├── change_hat
│   ├── change_hatv
│   ├── change_onexec
│   ├── change_profile
│   ├── computed_longest_left
│   ├── fix_binfmt_elf_mmap
│   ├── post_nnp_subset
│   ├── stack
│   └── version
├── file
│   └── mask
├── io_uring
│   └── mask
├── ipc
│   └── posix_mqueue
├── mount
│   └── mask
├── namespaces
│   ├── mask
│   ├── pivot_root
│   └── profile
├── network
│   ├── af_mask
│   └── af_unix
├── network_v8
│   └── af_mask
├── policy
│   ├── outofband
│   ├── permstable32
│   ├── permstable32_version
│   ├── set_load
│   ├── unconfined_restrictions
│   │   ├── change_profile
│   │   ├── io_uring
│   │   └── userns
│   └── versions
│       ├── v5
│       ├── v6
│       ├── v7
│       ├── v8
│       └── v9
├── ptrace
│   └── mask
├── query
│   └── label
│       ├── data
│       ├── multi_transaction
│       └── perms
├── rlimit
│   └── mask
└── signal
    └── mask
```
This is on kernel `6.5.0-21-generic #21-Ubuntu`, which should support prompting. From what I can see, kernel features are computed based on directories here. https://github.com/snapcore/snapd/blob/master/sandbox/apparmor/apparmor.go#L338

Is prompting covered by one of these features, or is it not yet listed in the supported kernel features?

Comment (alexmurray) [reply interfaces/builtin/home.go:115]: Apologies, I had assumed there was a particular feature file - instead we would need to parse `/sys/kernel/security/apparmor/features/policy/permstable32` and look for the string `prompt` within it - but perhaps it will be enough to just add support for probing this via `ParserFeatures()` - I am hoping that in the case that the kernel doesn't support the feature that the parser will raise an error.

Comment (olivercalder) [reply interfaces/builtin/home.go:115]: No worries, thanks for clarifying. I see `prompt` in `/v2/kernel/security/apparmor/features/policy/permstable32`, so that should be great, thank you. I'll look around if there's an existing helper function to parse and check this file.

I see that `ParserFeatures()` iterates through a list of features and checks each one with a probe string and optional flags. Do you know what these names/flags should be for prompting? I assume `prompt`, but not sure about any flags which might be required.

Comment (alexmurray) [reply interfaces/builtin/home.go:115]: So since `permstable32` is included in `KernelFeatures()` we may want to consider reading that file whilst parsing the existing features and exposing the values within `KernelFeatures()` somehow - then you could just check if a feature like `policy:perstable32:prompt` existing in `KernelFeatures()`.

For `ParserFeatures()`, I wouldn't worry about flags and just probe with a string like `prompt /foo r,` or somesuch - that should be sufficient I think.

**Comment by olivercalder** [comment]
Rebased on master to pull in recent changes, notably those from https://github.com/snapcore/snapd/pull/13693

File: interfaces/builtin/home.go [hunk 9]
```diff
@@ -85,13 +86,13 @@ audit deny @{HOME}/bin wl,
...
-@{HOME}/sna[^p]**      r,
-@{HOME}/snap[^/]**     r,
-@{HOME}/{s,sn,sna}{,/} r,
+###PROMPT###@{HOME}/               r,
+###PROMPT###@{HOME}/[^s.]**        r,
+###PROMPT###@{HOME}/s[^n]**        r,
+###PROMPT###@{HOME}/sn[^a]**       r,
+###PROMPT###@{HOME}/sna[^p]**      r,
+###PROMPT###@{HOME}/snap[^/]**     r,
+###PROMPT###@{HOME}/{s,sn,sna}{,/} r,
```
Comment (zyga) [inline interfaces/builtin/home.go:95]: Nitpick, perhaps we can just inject the prompt prefix to each non-empty, non-comment line?

File: interfaces/repo.go [hunk 10]
```diff
@@ -931,6 +931,14 @@ func (r *Repository) SnapSpecification(securitySystem SecuritySystem, appSet *Sn
 
 	spec := backend.NewSpecification(appSet)
 
+	if opts.AppArmorPrompting {
+		// Need to set UsePromptPrefix() here, but can't, because the backend
+		// might not be apparmor.
+		// And can't wait to do it in spec.AddConnectedPlug, since then it is
+		// re-added for every plug slot, which is innecessary.
+		// XXX: could do it in NewSpecification, but that is called ~750 times.
```
Comment (zyga) [inline interfaces/repo.go:939]: I remember adjusting tests when something changed in how we handle specifications and this had to be done for each of the interfaces and each test. Bazillion changes, can relate.

File: overlord/configstate/configcore/prompting.go [hunk 11]
```diff
@@ -0,0 +1,72 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//go:build !nomanagers
+// +build !nomanagers
```
Comment (zyga) [inline overlord/configstate/configcore/prompting.go:3]: Is the syntax with the space after the comment correct?

**Comment by zyga** [comment]
@olivercalder I have a feeling that enabling the experimental feature should require a restart of snapd. Enabling the value is easy. Dynamically doing this adds a lot of complexity that seems unwarranted. I'm not sure how you see this but I would focus on quickly getting to the point where we can see if the core is right, not if the interactions around it are nice.

We can always add this before this is non-experimental.

**Review by olivercalder — COMMENTED** [review]
Some notes (mostly for myself) from meeting with Samuele earlier today.

File: overlord/ifacestate/helpers.go [hunk 12]
```diff
@@ -206,7 +214,7 @@ func (m *InterfaceManager) regenerateAllSecurityProfiles(tm timings.Measurer) er
 			logger.Noticef("cannot get current info for snap %q: %s", snapName, err)
 			return interfaces.ConfinementOptions{}
 		}
-		opts, err := buildConfinementOptions(m.state, snapInfo, snapst.Flags)
+		opts, err := buildConfinementOptions(m.state, snapInfo, snapst.Flags, task)
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply overlord/ifacestate/helpers.go:217]: Since the task should be snap-specific, no need to pass `snapInfo` as well...?

File: overlord/ifacestate/handlers.go [hunk 13]
```diff
@@ -1960,3 +1973,15 @@ func (m *InterfaceManager) doHotplugSeqWait(task *state.Task, _ *tomb.Tomb) erro
 	// no conflicting change for same hotplug key found
 	return nil
 }
+
+func (m *InterfaceManager) doRegenerateAllSecurityProfiles(task *state.Task, _ *tomb.Tomb) error {
+	// XXX: this will make snapd unusable for a long time :(
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply overlord/ifacestate/handlers.go:1966]: Where do you think the task per snap should be created? AFAIU, `doRegenerateAllSecurityProfiles` is called when a `"regenerate-all-security-profiles"` task is added. Should `doRegenerateAllSecurityProfiles` create a task for each snap and have a handler something like `doRegenerateSnapSecurityProfile` (which would need to be added as well), or should `regenerateAllSecurityProfiles` create many tasks as it iterates through the snaps, or should a `"regenerate-all-security-profiles"` task *not* be used when the flag changes, and instead introduce a new kind of task to regenerate the profile for a single snap? What do you have in mind?

File: interfaces/builtin/home.go (hunk 5 above)
Comment (olivercalder) [reply interfaces/builtin/home.go]: Not strictly necessary for this PR, but what about `change_profile`, `io_uring`, and `userns` under `unconfined_restrictions`? Are those kernel features as well, and should they be included in the list? It seems like some features are not directory names -- sometimes file names (as those listed above), sometimes features listed in a file (e.g. permstable32`).

File: overlord/ifacestate/handlers.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (zyga) [reply overlord/ifacestate/handlers.go:1966]: I meant that the change could have a number of tasks like "setup-security" (existing task, existing handler) with the right snap setup info. I'm not sure if it works this way but at least we would have a shorter longest duration during which the lock is held and clearer conflict resolution.

File: interfaces/builtin/home.go (hunk 9 above)
Comment (olivercalder) [reply interfaces/builtin/home.go:95]: I believe there there will be some rules (perhaps in other interfaces rather than "home") for which we will not want to prompt. We could default to adding prompt with a `###NOPROMPT###` prefix to skip a given rule, but I prefer the current approach over that, I think.

File: overlord/configstate/configcore/prompting.go (hunk 7 above)
Comment (olivercalder) [reply overlord/configstate/configcore/prompting.go:33]: Changed in most recent commit, please let me know if this is what you had in mind.

File: overlord/configstate/configcore/prompting.go (hunk 11 above)
Comment (olivercalder) [reply overlord/configstate/configcore/prompting.go]: I think a typo from mvo, I've removed in most recent commit.

File: overlord/configstate/configcore/prompting.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	st.Unlock()
+	st.EnsureBefore(0)
+
+	select {
+	case <-regenerateProfilesChg.Ready():
+		st.Lock()
+		defer st.Unlock()
+		return regenerateProfilesChg.Err()
+	case <-time.After(10 * time.Minute):
+		// profile generate may take some time
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply overlord/configstate/configcore/prompting.go]: Thanks! With latest commit, introduced a ticker to log messages, but never hard timeout. Does this approach sound alright?

File: overlord/ifacestate/handlers.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (olivercalder) [reply overlord/ifacestate/handlers.go]: Thanks for the tip! Implemented in the latest commit. Please let me know if this is what you had in mind.

File: interfaces/builtin/home.go (hunk 5 above)
Comment (olivercalder) [reply interfaces/builtin/home.go]: Thanks for the suggestions! The latest commit has proper parser and kernel checking, and the feature is no longer exported.

File: interfaces/repo.go (hunk 10 above)
Comment (olivercalder) [reply interfaces/repo.go]: Aha, it's the backend-specific `NewSpecification` functions which are called ~750 times, the actual `Backend.NewSpecification` interface is only used a few dozen times :upside_down_face:

File: interfaces/seccomp/backend.go [hunk 15]
```diff
@@ -384,7 +384,7 @@ func generateContent(opts interfaces.ConfinementOptions, snippetForTag string, a
 }
 
 // NewSpecification returns an empty seccomp specification.
-func (b *Backend) NewSpecification(appSet *interfaces.SnapAppSet) interfaces.Specification {
+func (b *Backend) NewSpecification(appSet *interfaces.SnapAppSet, opts interfaces.ConfinementOptions) interfaces.Specification {
```
Comment (zyga) [inline interfaces/seccomp/backend.go:387]: Nitpick: This is named inconsistently in the code. I like "opts" more than "confinement," which is present in a few places above in this diff. Would you mind updating it all to "opts"?

File: overlord/configstate/configcore/prompting.go [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+	var prevPrompting bool
+	err = c.GetPristine(snap, confName, &prevPrompting)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if prompting == prevPrompting {
+		return nil
+	}
+
+	// XXX: what if apparmor-prompting flag value unchanged but support changed?
```
Comment (zyga) [inline overlord/configstate/configcore/prompting.go:52]: This makes me strongly believe that enabling prompting should require a restarting snapd. There is far less code to write, compexity to handle, and nobody cares about the tiny inconvenience involved.

File: overlord/configstate/configcore/prompting.go [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+	st.Lock()
+	regenerateProfilesChg := st.NewChange("regenerate-all-security-profiles",
+		i18n.G("Regenerate all profiles due to change in prompting"))
+	snapSetupProfileTasks, err := ifacestate.CreateSnapSetupProfilesTasks(st)
+	if err != nil {
+		return err
+	}
+	usePromptPrefix := prompting && features.AppArmorPrompting.IsSupported()
+	for _, t := range snapSetupProfileTasks {
+		t.Set("use-prompt-prefix", &usePromptPrefix)
```
Comment (zyga) [inline overlord/configstate/configcore/prompting.go:69]: Could we somehow inject this into snap setup profiles so that it lives in the same place as information about devmode or classic?

File: overlord/configstate/configcore/runwithstate.go [hunk 18]
```diff
@@ -66,6 +66,10 @@ func init() {
 
 	// kernel.{,dangerous-}cmdline-append
 	addWithStateHandler(validateCmdlineAppend, handleCmdlineAppend, &flags{modeenvOnlyConfig: true})
+
+	// experimental.apparmor-prompting
+	addWithStateHandler(nil, doExperimentalApparmorPromptProfileRegeneration, nil)
```
Comment (zyga) [inline overlord/configstate/configcore/runwithstate.go:71]: My earlier suggestion was to use the existing setup-snap-securoty task so that there is no new task type necessary. 

Is this infeasible? Do you want to sync and discuss this?

File: overlord/ifacestate/handlers.go [hunk 19]
```diff
@@ -81,20 +81,51 @@ func getExtraLayouts(st *state.State, snapInfo *snap.Info) ([]snap.Layout, error
...
-		ExtraLayouts: extraLayouts,
+		DevMode:           flags.DevMode,
+		JailMode:          flags.JailMode,
+		Classic:           flags.Classic,
+		ExtraLayouts:      extraLayouts,
+		AppArmorPrompting: usePromptPrefix(task),
 	}, nil
 }
 
+func buildDefaultUsePromptPrefixFunc(st *state.State) func(t *state.Task) bool {
```
Comment (zyga) [inline overlord/ifacestate/handlers.go:106]: Why is this required?

Could we just store the value and respect it when set, defaulting to off?

**Review by zyga — COMMENTED** [review]
Partial pass. I left some comments that may invalidate the rest of my review.

Please let me know if you want to sync on this. I'm trying to drive complexity down.

File: overlord/configstate/configcore/prompting.go (hunk 17 above)
Comment (olivercalder) [reply overlord/configstate/configcore/prompting.go:69]: The devmode and classic settings are in `confinementOptions` (are they elsewhere as well?). The PR adds a closure which we pass into `buildConfinementOptions`, and that (most of the time) checks to see if this value is set in the task. The reason to use the task is `buildConfinementOptions` might be called multiple times as part of the same change, and we want prompting to be consistently on or off for all confinement options related to that change.

But I agree, there's some redundancy here:
- If we don't define a custom closure, the default one will be created which:
  - first check if the task exists with the `"use-prompt-prefix"` value set, and if so, use that
  - otherwise, returns (supported && enabled), computed at time of closure creation (so it is consistent across invocations of the closure)
- If we do define a custom closure, we could have it ignore the task parameter and just return `true` or `false` based on some decision made at time of closure creation
  - However, we want to allow custom closures to have varying decisions based on the particular snap/task
  - This really is a niche use case though, I'm not sure when one would ever want to 

Basically, my current impulse is that optionally having both a task and/or a closure is redundant... My impulse is one would only ever have a task with the `"use-prompt-prefix"` value set if one has a custom closure, but then again, there could be a task created elsewhere when no closure is defined, and one *might* want to automatically cache the value on that task so the value is consistent with other `buildConfinementOptions` invocations with the same task.

**Review by olivercalder — COMMENTED** [review]
Very helpful sync with @zyga this morning -- thanks!

New (much simpler) approach:
- Embed prompting `supported && enabled` in system key, which triggers profile regeneration when the value changes.
- Don't regenerate profiles upon change in AppArmor Prompting feature flag until snapd restarts.
- Only run prompting listener if system key has prompting set.
- Use check for listener running (in memory, no state) to see if prompting prefix should be used.
- Unified `###PROMPT###` replacement for all interfaces as part of `interfaces/apparmor/backend.go`, akin to `###HOME_IX###` handling -- this is where to check if listener is running.

**Comment by pedronis** [comment]
> If flag is disabled on startup and enabled later:
System key will be set with prompting disabled, so listener will not start, so profile generation should never include prompt prefix
Importantly, we never want to include prompt prefix unless the listener is running, since otherwise application syscalls may block waiting for a prompt reply which will never arrive

@olivercalder what will restart snapd though when the flag is enabled in the UI?

**Comment by olivercalder** [comment]
To be clear, my most recent comment chain is a proposal for an alternate approach, rather than a description of this PR. I am working on implementing that alternative approach in a separate, parallel PR, so these approaches can be compared.

Additionally, in the meantime, I've opened #13821 to implement the AppArmor kernel and parser support check, as that is common between both approaches and orthogonal to the open questions addressed by this and the forthcoming PR.

File: interfaces/builtin/home.go (hunk 9 above)
Comment (olivercalder) [reply interfaces/builtin/home.go:95]: Ah, for that matter, these two rules are omitted from prompting:
```
# Disallow writes to the well-known directory included in
# the user's PATH on several distributions
audit deny @{HOME}/bin/{,**} wl,
audit deny @{HOME}/bin wl,
```
So we can't insert prompting for everything. We have to use something like `###PROMPT###` or `###NOPROMPT###`. I prefer the former.
